File: src/config/validation.rs
//! JSON-schema validation of the configuration document.
//!
//! The schema is **embedded** with `include_str!`, so it can never be "missing
//! from the classpath" — closing the fail-open hole in the Java validator. A
//! document that does not satisfy the schema is a hard error by default.

use serde_json::Value;

use crate::error::{GgError, Result};

const SCHEMA: &str = include_str!("../../resources/ggcommons-config-schema.json");

/// Validate `instance` against the embedded config schema.
pub fn validate(instance: &Value) -> Result<()> {
    let schema: Value = serde_json::from_str(SCHEMA)
        .map_err(|e| GgError::Validation(format!("embedded schema is not valid JSON: {e}")))?;

    let validator = jsonschema::validator_for(&schema)
        .map_err(|e| GgError::Validation(format!("embedded schema is invalid: {e}")))?;

    let errors: Vec<String> = validator
        .iter_errors(instance)
        .map(|e| format!("{e} (at {})", e.instance_path))
        .collect();

    if errors.is_empty() {
        Ok(())
    } else {
        Err(GgError::Validation(errors.join("; ")))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn accepts_a_well_formed_config() {
        let cfg = json!({
            "logging": { "level": "INFO" },
            "metricEmission": { "target": "cloudwatch", "namespace": "ns" },
            "heartbeat": { "intervalSecs": 5, "targets": [ { "type": "metric" } ] },
            "component": { "global": {}, "instances": [] }
        });
        assert!(validate(&cfg).is_ok());
    }

    #[test]
    fn rejects_a_bad_metric_target() {
        let cfg = json!({ "metricEmission": { "target": "not-a-target" } });
        assert!(validate(&cfg).is_err());
    }

    #[test]
    fn rejects_wrong_type_for_interval() {
        let cfg = json!({ "heartbeat": { "intervalSecs": "five" } });
        assert!(validate(&cfg).is_err());
    }
}

File: libs/ts/src/config/validation.ts
/**
 * Configuration — JSON-schema validation (fail-closed).
 *
 * The cross-language config schema is embedded ({@link ./schema.json}) so it can
 * never be "missing from the classpath" (closing the Java fail-open hole). A
 * validation failure is a hard error, mirroring the Rust/Python behavior. Uses
 * `ajv` (the standard JS JSON-schema validator).
 */
import Ajv, { ValidateFunction } from "ajv";

import { EdgeCommonsError } from "../errors";
import schema from "./schema.json";

let validator: ValidateFunction | undefined;

function compiled(): ValidateFunction {
  if (!validator) {
    const ajv = new Ajv({ allErrors: true, strict: false });
    validator = ajv.compile(schema as object);
  }
  return validator;
}

/**
 * Validate a raw config document against the embedded schema. Throws
 * {@link EdgeCommonsError} of kind `Validation` listing every error on failure.
 */
export function validate(instance: unknown): void {
  const v = compiled();
  if (!v(instance)) {
    const errors = (v.errors ?? [])
      .map((e) => `${e.instancePath || "(root)"} ${e.message}`)
      .join("; ");
    throw EdgeCommonsError.validation(`config failed schema validation: ${errors}`);
  }
}
